geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroCodecs.scala

```diff
@@ -0,0 +1,704 @@
+    def find(ct:ComplexType): AvroCodec[Property] = {
+      find(new AttributeDescriptorImpl(ct, ct.getName, 1, 1, false, null))
+    }
+
+    def find(ad:AttributeDescriptor): AvroCodec[Property] = {
+      codecs.getOrElseUpdate(ad, {
+        val schema = schemaFac(ad.getType.asInstanceOf[ComplexType])
+        val schemaType = getType(schema, fne)
+        val enc = bindEnc(schemaType, ad)
+        val dec = bindDec(schemaType, ad)
+        bind[Property](enc,dec,schema)
+      })
+    }
+
+    private[this] def bindEnc(s:SchemaType, ad:PropertyDescriptor):EncodeProperty = s match {
+      case r@Record(fields) if ad.getType.isInstanceOf[ComplexType] =>
+        val ct = ad.getType.asInstanceOf[ComplexType]
+      encoders.get(r) match {
+        case Some(enc) => return enc
+        case _ =>
+      }
+
+      if(constructing(r)) {
+        return new EncodeProperty {
+          private[this] lazy val binding=encoders(r)
+          override def apply(e: Encoder, p: Property): Unit = binding(e,p)
+        }
+      }
+
+      constructing(r) = true
+
+      val descByName = AvroSchemas.allProps(ct).map(d => d.getName -> d).toMap
+
+      // Each field MUST be mapped
+      val bindings = r.fields.map { case (name, st) =>
+        if (AvroSchemas.AVRO_NAMESPACE == name.getNamespaceURI) {
+          name.getLocalPart match {
+            case "__version__" =>
+              new EncodeComplexAttribute {
+                def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) = e.writeInt(1)
+              }
+
+            case "__fid__" =>
+              new EncodeComplexAttribute {
+                def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) =
+                  e.writeString(Option(ca.getIdentifier) flatMap (i => Option(i.getID)) map (_.toString) getOrElse "")
+              }
+
+            case "simpleContent" => st match {
+              case n@Nullable(Primitive(t)) =>
+                val binding = bindEncPrimitive(AvroSchemas.nearestSimpleType(ct), t)
+                new EncodeComplexAttribute {
+                  def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) =
+                    Option(ca.getProperty(GEOT_SIMPLE_CONTENT)) match {
+                      case Some(value) if value.getValue != null =>
+                        n.writeNonNull(e)
+                        binding(e, value)
+                      case _ =>
+                        n.writeNull(e)
+                    }
+                }
+
+              case _ =>
+                throw new IllegalStateException(s"The property $name must be a nullable primitive type")
+            }
+
+            case other => st match {
+              case n: Nullable =>
+                new EncodeComplexAttribute {
+                  def apply(ca: ComplexAttribute, p: JHashMap[Name, JArrayList[Property]], e: Encoder) = n.writeNull(e)
+                }
+              case _ => ???

```

Comment:
Ok, how can one get here?  

Can throw an exception which might be helpful to the user?